### ISSUE TITLE
The log shows - "ClientClosedError: The client is closed" even having explicit connect & disconnect for each call

### ISSUE THREAD

**Issue by billsenxu** [redis/node-redis#2607]
### Description

Sometime the node.js application would throw out the error message - "ClientClosedError: The client is closed" even having explicit connect & disconnect for each call in the code. As redis is used frequenty in the program, the redis is initiated when starting the program(with built-in createClient() method). In my scenario of every express call(get or post), the redis connect will be called if "redisClient.isOpen" is false, after the redis access, the redis disconnect will be excuted explicitly. Don't  know why this error was caused.

By the way, if the redis diconnect method is unnecessary(as the front-end users will access the program from time to time, so logically the redis would disconnect and connect alternatively all along)?

Appreciate for the kind assistance.

### Node.js Version

v18.14.2

### Redis Server Version

7.2.0

### Node Redis Version

redis@4.6.7

### Platform

Ubuntu 20.04(Linux5.4.0-148-genericx86_64)

### Logs

```bash
/home/abc/node_modules/@redis/client/dist/lib/client/index.js:490
        return Promise.reject(new errors_1.ClientClosedError());
                              ^

ClientClosedError: The client is closed
    at Commander._RedisClient_sendCommand (/home/abc/node_modules/@redis/client/dist/lib/client/index.js:490:31)
    at Commander.commandsExecutor (/home/abc/node_modules/@redis/client/dist/lib/client/index.js:188:154)
    at BaseClass.<computed> [as rPush] (/home/abc/node_modules/@redis/client/dist/lib/commander.js:8:29)
    at /home/abc/Frontend.js:870:29
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by adadgio** [comment]
Any update on this issue? Still happening...

**Comment by leibale** [comment]
You shouldn't `connect` and `disconnect` every time you want to run a command, you should `connect` when the application starts and `disconnect` when you want to gracefully close the application (if you ever need to).

Assuming you are using it with an HTTP server, if you `connect` and `disconnect` in a request handler, when there are 2 requests in parallel, whenever the first request finishes, it'll close the connection for the second request as well...

Hope that makes sense..

**Comment by billsenxu** [comment]
As recommended by  leibale, no need to disconnect for every http request, as the program normally is running all along. The reason I executed the "disconnect" previously  for each request was trying to prevent the possible more and more open connections, based on the observation, this case wouldn't happen.

**Comment by radirantast** [comment]
> You shouldn't `connect` and `disconnect` every time you want to run a command, you should `connect` when the application starts and `disconnect` when you want to gracefully close the application (if you ever need to).
> 
> Assuming you are using it with an HTTP server, if you `connect` and `disconnect` in a request handler, when there are 2 requests in parallel, whenever the first request finishes, it'll close the connection for the second request as well...
> 
> Hope that makes sense..

actually that sounds like a great idea, but how I can do this?

**Comment by leibale** [comment]
@radirantast

redis.js:
```javascript
import { createClient } from 'redis';

export const client = await createClient()
  .on('error', err => console.error('Redis Client Error', err))
  .connect();
```

index.js:
```javascript
import { createServer } form 'node:http';
import { client } from './redis.js';

const server = createServer(async (req, res) => {
  try {
    res.end(await client.get('key'));
  } catch (err) {
    console.error(err);
    res.end('Internal Server Error');
  }
});

server.listen(3000);
```

**Comment by noodles-1** [comment]
> Collaborator

This is what I have done in my Next.js project, but for some reason, it still gives the same error "ClientClosedError: The client is closed" when fetch onto the server that runs redis commands. I have looked at all my files and there is no `await client.disconnect()` nor `await client.quit()`.

**Comment by noodles-1** [comment]
> You shouldn't `connect` and `disconnect` every time you want to run a command, you should `connect` when the application starts and `disconnect` when you want to gracefully close the application (if you ever need to).
> 
> Assuming you are using it with an HTTP server, if you `connect` and `disconnect` in a request handler, when there are 2 requests in parallel, whenever the first request finishes, it'll close the connection for the second request as well...
> 
> Hope that makes sense..

This is what I have done in my Next.js project, but for some reason, it still gives the same error "ClientClosedError: The client is closed" when fetch onto the server that runs redis commands. I have looked at all my files and there is no `await client.disconnect()` nor `await client.quit()`.

**Comment by joshpwrk** [comment]
running into the same exact issue. 

This is how the client connection is made

```
export async function getRedisCluster(hostEnvName: string, portEnvName: string): Promise<RedisCluster> {
  const args = getRedisArgs(hostEnvName, portEnvName);

  logger.info(`connecting to Redis cluster at: ${args.redisHost}:${args.redisPort}...`);

  const redisCluster = redis.createCluster({
    rootNodes: [
      {
        url: `redis://${args.redisHost}:${+args.redisPort}`,
      },
    ],
    // minimizeConnections: true,
  });

  redisCluster.on('error', (err) => console.log('Redis Cluster Error', err));

  await redisCluster.connect();

  return redisCluster;
}

export const subscriptionClient = await getRedisCluster('REDIS_SUBSCRIPTION_HOST', 'REDIS_SUBSCRIPTION_PORT');
```

This is an example of how the unsubscribe happens.
```
export async function maybeRemoveGlobalSubscription(channel: string) {
  liveWebSockets.forEach((wsc) => {
    if (wsc.subscribedChannels?.includes(channel)) {
      return; // early return if any other client is still subscribed
    }
  });

  if (globalSubscriptions.has(channel)) {
    globalSubscriptions.delete(channel);

    await subscriptionClient.sUnsubscribe(channel);
  }
}
```

And this is the error I get
```
Error: The client is closed
    at RedisSocket.disconnect (/app/node_modules/@redis/client/dist/lib/client/socket.js:63:19)
    at Commander.disconnect (/app/node_modules/@redis/client/dist/lib/client/index.js:343:64)
    at RedisClusterSlots.executeShardedUnsubscribeCommand (/app/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:154:26)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async maybeRemoveGlobalSubscription (file:///app/build/routes/subscribe/relayer.js:16:9)
    at async file:///app/build/routes/subscribe/relayer.js:5:9 [/app/build/utils/errors.js]
ErrorMessage: The client is closed
StackTrace: Error: The client is closed
    at RedisSocket.disconnect (/app/node_modules/@redis/client/dist/lib/client/socket.js:63:19)
    at Commander.disconnect (/app/node_modules/@redis/client/dist/lib/client/index.js:343:64)
    at RedisClusterSlots.executeShardedUnsubscribeCommand (/app/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:154:26)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async maybeRemoveGlobalSubscription (file:///app/build/routes/subscribe/relayer.js:16:9)
    at async file:///app/build/routes/subscribe/relayer.js:5:9
```

**Comment by Fatimaarshad10** [comment]
I encountered an issue with the **redis** library when using version 4, but switching to version 3.1 resolved the problem. It seems there might be compatibility or changes in behavior between versions 4 and 3.1 that affected my application. I wanted to share this information in case others face a similar issue and might find it helpful to try using version 3.1 instead

**Comment by noodles-1** [comment]
This is what I've done on my Node/Express project:

redis.js:
```javascript
import { createClient } from 'redis';

export const client = createClient({
    url: process.env.REDIS_URL,
})

if (!client.isOpen) {
    await client.connect()
    console.log('Connected to Redis')
}
``` 

```javascript
import 'dotenv/config'

import express from 'express'
import bodyParser from 'body-parser'
import cors from 'cors'
import { router } from './api.js'

const corsConfig = {
    origin: 'http://localhost:3000'
}

const app = express()
const PORT = process.env.PORT

app.use(cors(corsConfig))
app.use(bodyParser.json())
app.use('/api', router)

app.listen(PORT, () => console.log(`Listening to port ${PORT}`))
```


In api.js, I'm simply importing functions from my actions directory such as in my case, updating a note:

updateNote.js:
```javascript
import { client } from "../redis.js"
import { noteSchema } from "../schemas/index.js"
import { Repository } from "redis-om"

export async function updateNote(note) {
    const repository = new Repository(noteSchema, client)
    const result = await repository.fetch(note.noteId)

    result.title = note.title
    result.body = note.body
    result.category = note.category
    result.modified = note.modified
    await repository.save(result)
}
```

When running `npm run start`, it automatically connects to Redis and preserves that connection.

**Comment by noodles-1** [comment]
This is what I've done on my Node/Express project:

redis.js:
```javascript
import { createClient } from 'redis';

export const client = createClient({
    url: process.env.REDIS_URL,
})

if (!client.isOpen) {
    await client.connect()
    console.log('Connected to Redis')
}
``` 

index.js:
```javascript
import 'dotenv/config'

import express from 'express'
import bodyParser from 'body-parser'
import cors from 'cors'
import { router } from './api.js'

const corsConfig = {
    origin: process.env.CORS_ORIGIN
}

const app = express()
const PORT = process.env.PORT

app.use(cors(corsConfig))
app.use(bodyParser.json())
app.use('/api', router)

app.listen(PORT, () => console.log(`Listening to port ${PORT}`))
```


In api.js, I'm simply importing functions from my actions directory such as in my case, updating a note:

updateNote.js:
```javascript
import { client } from "../redis.js"
import { noteSchema } from "../schemas/index.js"
import { Repository } from "redis-om"

export async function updateNote(note) {
    const repository = new Repository(noteSchema, client)
    const result = await repository.fetch(note.noteId)

    result.title = note.title
    result.body = note.body
    result.category = note.category
    result.modified = note.modified
    await repository.save(result)
}
```

When running `npm run start`, it automatically connects to Redis and preserves that connection.

**Comment by sahanmndl** [comment]
@noodles-1 Thanks for your code, I have fixed the connection error. I wanted to ask are you using `redis-om` for caching the API response? I am using the normal `redisClient.setEx()` from the `redis` library, but upon sending a API request, it just shows sending request indefinitely, without any error or result. I'm using Redis cloud database.

**Comment by BrentLayne** [comment]
@joshpwrk I've opened a new issue because I'm running into the same problem as you and I've definitely found a bug with how `SUNSUBSCRIBE` works: https://github.com/redis/node-redis/issues/2685

**Comment by aleksandregvrm** [comment]
### whenever i run await client.connect() in the controller, the redis server works fine  but if i try to to use client.get() method in the mongoose.exec i get the above mentioned error . if i try to use client.connect() in the exec function i get this error Error: Socket already opened. what should i do? 



`const Product = require("./ProductModel");
const redis = require("redis");
const url = "redis://127.0.0.1:6379";

const getAllProducts = async (req, res) => {
    const client = redis.createClient(url);
    await client.connect()
    const cached = await client.get('countries',10);
    // const cachedCountries = JSON.parse(cached);
    // // if(cachedCountries !== null){
    // //     console.log('getting from cached');
    // //     return res.status(200).json(cachedCountries)
    // // }
    const products = await Product.find({productAmount:12}).cache();
    await client.set('countries',JSON.stringify(products));
    console.log('getting from mongo');
    res.status(200).json(products);
};

const setProduct = async (req, res) => {
    const product = await Product.create(req.body);
    res.status(200).json(product);
};

module.exports = { getAllProducts, setProduct };
`
 `const redis = require("redis");
const mongoose = require("mongoose");
const exec = mongoose.Query.prototype.exec;
const url = "redis://127.0.0.1:6379";
const client = redis.createClient(url);

mongoose.Query.prototype.cache = function (options = {}) {
  this.useCache = true;
  return this;
};
mongoose.Query.prototype.exec = async function () {
  if (!this.useCache) {
    return exec.apply(this, arguments);
  }
  await client.connect();
  const cachedData = JSON.parse(await client.get("countries"));
  console.log(cachedData);
  const key = Object.assign({}, this.getQuery(), {
    collection: this.mongooseCollection.name,
  });
  console.log(key);
  return exec.apply(this, arguments);
};`

**Comment by paramvirTheBlockCoders** [comment]
please use this

const redis = require('redis'); 
const client = redis.createClient();

client.connect(); 
client.on('connect', async function() { 
    console.log('Connected!'); 
});

**Comment by RR1335** [comment]
const redis = require('ioredis') , Done.
